Importing numpy library as np.

In [1]:
import numpy as np

Creating coefficient matrix and result matrix.

In [2]:
A = np.array([
    [10.0, 2.0, -1.0],
    [-3.0, -6.0, 2.0],
    [1.0, 1.0, 5.0]
])

B = np.array([27.0, -61.5, 21.5])

Solving the linear system directly using numpy methods. (Gauss-Jordan)

In [17]:
x_solve = np.linalg.solve(A, B)
print(f"Gauss-Jordan results:")
print(f"x1 = {x_solve[0]}")
print(f"x2 = {x_solve[1]}")
print(f"x3 = {x_solve[2]}")

Gauss-Jordan results:
x1 = 0.7975778546712803
x2 = 10.529411764705882
x3 = 2.0346020761245676


Implementing Gauss-Seidel iteration. (5 Iteration)

In [18]:
# Starting values for x1, x2 and x3
x = np.array([0.0, 0.0, 0.0])

# Iteration count
num_iterations = 5

# Table for results
print("k |   x1       |   x2       |   x3")
print(f"0 | {x[0]:.8f} | {x[1]:.8f} | {x[2]:.8f}")

# Iteration loop (5 Loop)
for k in range(num_iterations):
    # x_new will store current iteration's values.
    x_new = np.array([0.0, 0.0, 0.0])

    # Founding x1 value using formula.
    x_new[0] = (27.0 - 2.0 * x[1] + 1.0 * x[2]) / 10.0

    # Founding x2 value using formula.
    x_new[1] = (-61.5 + 3.0 * x_new[0] - 2.0 * x[2]) / -6.0

    # Founding x3 value using formula.
    x_new[2] = (21.5 - x_new[0] - x_new[1]) / 5.0

    # Updating x value for next iteration
    x = x_new

    # Printing current iteration results
    print(f"{k+1} | {x[0]:.8f} | {x[1]:.8f} | {x[2]:.8f}")

print("\nGauss-Seidel 5th iteration results:")
print(f"x1 = {x[0]:.8f}")
print(f"x2 = {x[1]:.8f}")
print(f"x3 = {x[2]:.8f}")

k |   x1       |   x2       |   x3
0 | 0.00000000 | 0.00000000 | 0.00000000
1 | 2.70000000 | 8.90000000 | 1.98000000
2 | 1.11800000 | 10.35100000 | 2.00620000
3 | 0.83042000 | 10.50352333 | 2.03321133
4 | 0.80261647 | 10.52642888 | 2.03419093
5 | 0.79813332 | 10.52899698 | 2.03457394

Gauss-Seidel 5th iteration results:
x1 = 0.79813332
x2 = 10.52899698
x3 = 2.03457394


Gauss-Jordan results:

x1 = 0.7975778546712803

x2 = 10.529411764705882

x3 = 2.0346020761245676

Gauss-Seidel 5th iteration results:

x1 = 0.79813332

x2 = 10.52899698

x3 = 2.03457394

As we can see our solutions are very close.